File: phase-2/frontend/lib/types/task.ts
/**
 * TypeScript types for Task entity.
 *
 * Matches backend Task SQLModel schema from:
 * phase-2/backend/src/models/task.py
 *
 * @see specs/001-fullstack-todo-web/spec.md - Data Structures
 */

/**
 * Task status enum - matches backend TaskStatus.
 */
export type TaskStatus = "pending" | "complete";

/**
 * Task priority enum - matches backend TaskPriority.
 */
export type TaskPriority = "Low" | "Medium" | "High" | "Urgent";

/**
 * Task entity - full task representation.
 *
 * Matches TaskRead schema from backend.
 */
export interface Task {
  /** Unique identifier (UUID4) */
  id: string;
  /** Owner user ID (UUID4) */
  user_id: string;
  /** Task title (1-200 characters) */
  title: string;
  /** Task description (0-2000 characters, nullable) */
  description: string | null;
  /** Task status: pending or complete */
  status: TaskStatus;
  /** Task priority: Low, Medium, High, or Urgent */
  priority: TaskPriority;
  /** Optional category ID (UUID4) */
  category_id: string | null;
  /** Creation timestamp (ISO 8601) */
  created_at: string;
  /** Last modification timestamp (ISO 8601) */
  updated_at: string;
  /** Completion timestamp (ISO 8601, nullable) */
  completed_at: string | null;
}

/**
 * Task creation payload.
 *
 * Matches TaskCreate schema from backend.
 */
export interface TaskCreate {
  /** Task title (required, 1-200 characters) */
  title: string;
  /** Task description (optional, max 2000 characters) */
  description?: string | null;
  /** Task priority (default: Medium) */
  priority?: TaskPriority;
  /** Category ID to assign (optional) */
  category_id?: string | null;
  /** Tag IDs to assign (optional) */
  tag_ids?: string[];
}

/**
 * Task update payload.
 *
 * Matches TaskUpdate schema from backend.
 * All fields are optional - only provided fields are updated.
 */
export interface TaskUpdate {
  /** Updated title */
  title?: string;
  /** Updated description */
  description?: string | null;
  /** Updated priority */
  priority?: TaskPriority;
  /** Updated category ID */
  category_id?: string | null;
  /** Updated tag IDs */
  tag_ids?: string[];
}

/**
 * Task completion toggle payload.
 *
 * Matches TaskComplete schema from backend.
 */
export interface TaskComplete {
  /** New status (pending or complete) */
  status: TaskStatus;
}

/**
 * Task list response from backend.
 */
export interface TaskListResponse {
  /** List of tasks */
  tasks: Task[];
  /** Total count of tasks matching filters */
  total: number;
}

/**
 * Task filter parameters for list endpoint.
 *
 * T131-T134: Enhanced filters for task discovery.
 * Note: Using explicit `| undefined` for exactOptionalPropertyTypes compatibility.
 */
export interface TaskFilters {
  /** Filter by status (pending, complete, all) */
  status?: TaskStatus | "all" | undefined;
  /** Filter by priority (comma-separated for multiple) */
  priority?: string | undefined;
  /** Filter by category ID */
  category?: string | undefined;
  /** Filter by tag IDs (comma-separated, AND logic) */
  tags?: string | undefined;
  /** Search query for title/description */
  search?: string | undefined;
  /** Filter tasks created after this date (ISO 8601) */
  created_after?: string | undefined;
  /** Filter tasks created before this date (ISO 8601) */
  created_before?: string | undefined;
  /** Filter tasks updated after this date (ISO 8601) */
  updated_after?: string | undefined;
  /** Filter tasks updated before this date (ISO 8601) */
  updated_before?: string | undefined;
  /** Sort field */
  sort_by?: "created_at" | "updated_at" | "title" | "priority" | undefined;
  /** Sort direction */
  order?: "asc" | "desc" | undefined;
  /** Maximum results per page */
  limit?: number | undefined;
  /** Number of results to skip */
  offset?: number | undefined;
}

/**
 * Task with expanded relationships.
 *
 * Used when fetching task with category and tags loaded.
 */
export interface TaskWithRelations extends Task {
  /** Expanded category object */
  category?: {
    id: string;
    name: string;
    color: string | null;
  } | null;
  /** List of associated tags */
  tags?: Array<{
    id: string;
    name: string;
    color: string | null;
  }>;
}

File: phase-3/backend/src/models/task.py
"""Task SQLModel - Extended from Phase I with multi-user support and organization features."""

from datetime import datetime
from typing import Optional
from uuid import UUID, uuid4
import sqlalchemy as sa
from sqlmodel import Field, SQLModel, Column
from enum import Enum


class TaskStatus(str, Enum):
    """Task status enumeration."""
    PENDING = "pending"
    COMPLETE = "complete"


class TaskPriority(str, Enum):
    """Task priority enumeration."""
    LOW = "Low"
    MEDIUM = "Medium"
    HIGH = "High"
    URGENT = "Urgent"


class Task(SQLModel, table=True):
    """Task entity - Extended from Phase I with multi-user support."""

    __tablename__ = "tasks"

    # Phase I attributes (preserved)
    id: UUID = Field(default_factory=uuid4, primary_key=True, nullable=False)
    title: str = Field(max_length=200, nullable=False)
    description: Optional[str] = Field(default=None, max_length=2000)
    status: TaskStatus = Field(default=TaskStatus.PENDING, nullable=False)
    created_at: datetime = Field(default_factory=datetime.utcnow, nullable=False, index=True)
    updated_at: datetime = Field(default_factory=datetime.utcnow, nullable=False)
    completed_at: Optional[datetime] = Field(default=None)
    due_date: Optional[datetime] = Field(default=None)

    # Phase II new attributes
    user_id: str = Field(sa_column=Column(sa.VARCHAR(64), nullable=False, index=True))
    priority: TaskPriority = Field(default=TaskPriority.MEDIUM, nullable=False, index=True)
    category_id: Optional[UUID] = Field(default=None, foreign_key="categories.id", index=True)


class TaskCreate(SQLModel):
    """Schema for creating a task."""
    title: str
    description: Optional[str] = None
    status: TaskStatus = TaskStatus.PENDING
    priority: TaskPriority = TaskPriority.MEDIUM
    category_id: Optional[UUID] = None
    tag_ids: Optional[list[UUID]] = None
    user_id: Optional[str] = None


class TaskRead(SQLModel):
    """Schema for task responses."""
    id: UUID
    user_id: str
    title: str
    description: Optional[str]
    status: TaskStatus
    priority: TaskPriority
    category_id: Optional[UUID]
    created_at: datetime
    updated_at: datetime
    completed_at: Optional[datetime]
    due_date: Optional[datetime] = None


class TaskUpdate(SQLModel):
    """Schema for updating a task."""
    title: Optional[str] = None
    description: Optional[str] = None
    priority: Optional[TaskPriority] = None
    category_id: Optional[UUID] = None
    tag_ids: Optional[list[UUID]] = None


class TaskComplete(SQLModel):
    """Schema for toggling task completion status."""
    status: TaskStatus
